### ISSUE TITLE
Twitter and Tumblr 2FA code is never correct

### ISSUE THREAD

**Issue by Soldjermon** [winauth/winauth#449]
So, I enter in my credentials to login to both **Twitter** and **Tumblr**, everything is correct, so no worries there.

Here's the deal of what I'm encountering.

I've done the sync time, had correct secret key, and it generates the 2FA codes fine on **winauth**.

But! Whenever I try to log into either **Twitter** or **Tumblr** and try to use the *2FA* code from **winauth**, I get an error saying that the *2FA* code is incorrect from both websites the browser I'm using.

Though, I'm lucky that the *2FA* codes still send to my phone to log in and they work fine.

Heck, I've even regenerated with *new* secret keys and tried to use them on **winauth** to login and such, it still gives the same error from both websites.

I hope this'll help in fixing the issue if it is the case of **winauth** itself.

- Soldjermon

**Comment by winauth** [comment]
Are you saying your phone and WinAuth with the same keys are giving different codes?

Put the secret key into Google Authenticator on you phone. Does it have the same code as WinAuth?

**Comment by Soldjermon** [comment]
Hmm... Well, I've just tried with Authy with Tumblr, it has the same problem of not recognizing the code. So, I'm guessing it's a setting in the two websites that I have turned on or something, which sends the code to my SMS text instead and not through *any* authenticator.

Comparing the SMS text message and the authenticator codes, they're *completely* different.
Though, the authenticator codes between Authy and Winauth are the same(since using the same secret key).

I'm really sorry, it could be that I'm dumb... Can it be from Twitter, "Verify login requests" turned on in the settings, would not let the authenticators to not work properly for Twitter and only send as SMS text message?

Also, I don't *exactly* have a phone(can't afford one)... so, I use google voice to get SMS text messages, on a different, more secure computer; so using a google authenticator app on a real phone is out of the question.

**Comment by winauth** [comment]
SMS and authenticator codes are completely different. It depends on how Tumblr/Twitter implemented them, but I've never expected them to be the same. They use different keys. 

If you use SMS authenticator, you can't use WinAuth. You would have to change your login on that account to use the type of authenticator where you add the key into something like Authy, Google Authenticator or WinAuth.

It looks like Tumblr, only supports SMS validation so WinAuth can't be used. It's unfortunate that they only do this, especially since NIST recommend against SMS authenticator due to its potential vulnerabilities.

**Comment by Soldjermon** [comment]
Yeah... I'm sorry about this. :( I didn't realize that Twitter/Tumblr isn't allowing Winauth, Authy or any other authenticator(on PCs at least) to be used; even though they work correctly on them with the secret keys(from generating an unique app password) from both websites and use them on the authenticators. But, they just don't accept the code at all from the authenticators and say that the code is always incorrect.

So, it saddens me that I'm not allowed to use any authenticator on the PC on the two websites.

**Comment by My1** [comment]
the app password has nothing to do with the OTP-seed/secret.,
the app password usually is just a bunch of letters meaning that they fit into the base32 character set which coincidentially is what is used to transfer the OTP Seed.
but the OTP seed is just the base for generating an OTP, if the other side isnt trying for TOTP/HOTP authorization (meaning they use the same seed to generate an OTP for comparison) it's never going to work.

**Comment by codemasher** [comment]
While using WinAuth to verify my own code, i've just noticed an off-by-one error for counter based codes, which may be the answer to this issue.

![image](https://user-images.githubusercontent.com/592497/34311704-51a1cf80-e75f-11e7-8b6a-796713e16b13.png)

**Comment by My1** [comment]
actually TOTP is HOTP but with the time instead of the counter everything works identiallcy. if the time counter (unix time divided by 30) would be off by 1 too that might already throw off some services that dont have a tolerance window.

this could easily be tested. put the same secret into winauth and something else (e.g. google authenticator on a phone), make sure the times are identical and check the codes.

**Comment by codemasher** [comment]
https://github.com/codemasher/php-authenticator/blob/master/tests/AuthenticatorTest.php#L233-L322
